js/modules/loki_public_chat_api.js

```diff
@@ -72,13 +79,102 @@ class LokiPublicServerAPI {
     this.channels.splice(i, 1);
     thisChannel.stopPolling = true;
   }
+
+  async getServerToken() {
+    let token = await Signal.Data.getPublicServerTokenByServerName(this.server);
+    if (!token) {
+      token = await this.getNewToken();
+      if (token) {
+        await Signal.Data.savePublicServerToken({
+          server: this.server,
+          token,
+        });
+      }
+    }
+    return token;
+  }
+
+  async getNewToken() {
+    if (!this.tokenPending) {
+      this.tokenPending = true;
+      this.tokenPromise = new Promise(async res => {
+        const token = await this.requestToken();
+        if (!token) {
+          res(null);
+          return;
+        }
+        const registered = await this.submitToken(token);
+        if (!registered) {
+          res(null);
+          return;
+        }
+        res(token);
+      });
+    }
+    const token = await this.tokenPromise;
+    this.tokenPending = false;
+    return token;
+  }
+
+  async requestToken() {
+    const url = new URL(`${this.baseServerUrl}/loki/v1/get_challenge`);
+    const params = {
+      pubKey: this.chatAPI.ourKey,
+    };
+    url.search = new URLSearchParams(params);
+
+    let res;
+    try {
+      res = await nodeFetch(url);
+    } catch (e) {
+      return null;
+    }
+    if (!res.ok) {
+      return null;
+    }
+    const body = await res.json();
+    const { cipherText64, serverPubKey64 } = body;
+    const token = await libloki.crypto.decryptToken(
+      cipherText64,
+      serverPubKey64
```

Comment:
Style nit: if you changed `decryptToken` to take an object `{ cipherText64, serverPubKey64}` instead of the two arguments separately:
- you reduce the risk of passing variables in the wrong order
- you wouldn't need to deconstruct body, you could just pass it directly to `decryptToken`

Not super important tho, maybe there are cons for doing this?

Reply:
Yep I think it's better

Reply:
> parse time instead of run time
Interesting, can you elaborate on how you check that all the variables are passed (genuine question)? Is it done by a linter?